File: src/de/hhu/bsinfo/utils/StringUtils.java
/*
 * Copyright (C) 2017 Heinrich-Heine-Universitaet Duesseldorf, Institute of Computer Science, Department Operating Systems
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

package de.hhu.bsinfo.utils;

import java.io.UnsupportedEncodingException;
import java.util.regex.Pattern;

/**
 * Various string related utility functions.
 *
 * @author Florian Klein, florian.klein@hhu.de, 05.02.2014
 * @author Stefan Nothaas, stefan.nothaas@hhu.de, 31.03.2017
 */
public final class StringUtils {

    /**
     * Utils class
     */
    private StringUtils() {

    }

    /**
     * Checks if a Byte array contains a String (as much precise as possible)
     *
     * @param p_array
     *         the byte array
     * @return whether the byte array contains a String
     * @throws UnsupportedEncodingException
     *         if the encoding is unsupported
     */
    public static boolean looksLikeUTF8(final byte[] p_array) throws UnsupportedEncodingException {
        String phonyString;
        Pattern p;

        p = Pattern.compile(
                "\\A(\n" + "  [\\x09\\x0A\\x0D\\x20-\\x7E]             # ASCII\\n" + "| [\\xC2-\\xDF][\\x80-\\xBF]               # non-overlong 2-byte\n" +
                        "|  \\xE0[\\xA0-\\xBF][\\x80-\\xBF]         # excluding overlongs\n" +
                        "| [\\xE1-\\xEC\\xEE\\xEF][\\x80-\\xBF]{2}  # straight 3-byte\n" +
                        "|  \\xED[\\x80-\\x9F][\\x80-\\xBF]         # excluding surrogates\n" + "|  \\xF0[\\x90-\\xBF][\\x80-\\xBF]{2}      # planes 1-3\n" +
                        "| [\\xF1-\\xF3][\\x80-\\xBF]{3}            # planes 4-15\n" + "|  \\xF4[\\x80-\\x8F][\\x80-\\xBF]{2}      # plane 16\n" + ")*\\z",
                Pattern.COMMENTS);

        phonyString = new String(p_array, "UTF-8");

        return p.matcher(phonyString).matches();
    }
}

File: ibdxnet/src/ibnet/sys/StringUtils.cpp
/*
 * Copyright (C) 2017 Heinrich-Heine-Universitaet Duesseldorf,
 * Institute of Computer Science, Department Operating Systems
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

#include "StringUtils.h"

#include <sstream>

namespace ibnet {
namespace sys {

std::vector<std::string> StringUtils::Split(const std::string& text,
        const std::string& delimiter, bool ignoreEmptyTokens)
{
    std::string remaining = text;
    std::vector<std::string> result;
    std::string::size_type pos = remaining.find(delimiter);

    while (pos != std::string::npos)
    {
        // drop 0 sized strings
        std::string substr = remaining.substr(0, pos);
        if (substr.size() != 0 || !ignoreEmptyTokens)
        {
            result.push_back(substr);
        }

        remaining = remaining.substr(pos + 1);
        pos = remaining.find(delimiter);

        // using g++ the while condition was ignored for one
        // case though pos was set to npos (wtf)
        // adding this and everything's good again...
        if (pos == std::string::npos)
            break;
    }

    if (remaining.size() != 0)
        result.push_back(remaining);

    return result;
}

std::string StringUtils::ToHexString(uint64_t value, bool fillZeros, bool hexNumberIdent)
{
    return __ToHexString(value, fillZeros ? 16 : 0, hexNumberIdent);
}

std::string StringUtils::ToHexString(uint32_t value, bool fillZeros, bool hexNumberIdent)
{
    return __ToHexString(value, fillZeros ? 8 : 0, hexNumberIdent);
}

std::string StringUtils::ToHexString(uint16_t value, bool fillZeros, bool hexNumberIdent)
{
    return __ToHexString(value, fillZeros ? 4 : 0, hexNumberIdent);
}

std::string StringUtils::ToHexString(uint8_t value, bool fillZeros, bool hexNumberIdent)
{
    return __ToHexString(value, fillZeros ? 2 : 0, hexNumberIdent);
}

std::string StringUtils::__ToHexString(uint64_t value, uint32_t fillZerosCount, bool hexNumberIdent)
{
    std::ostringstream os;
    os << std::hex << value;

    std::string tmp;

    if (hexNumberIdent)
        tmp += "0x";
    if (fillZerosCount > 0)
    {
        for (size_t i = os.str().size(); i < fillZerosCount; i++)
            tmp += "0";
    }

    tmp += os.str();

    return tmp;
}

}
}